### PR TITLE
audio: Sync animation with embedded audio streams

### LINKED ISSUE #3020: Wrong frame rate for swf loaded via  GetUrl2
When playing the page http://www.hwjyw.com/textbook/dzjc/zh/cd3/main4/index.htm dialogue in Firefox, video and audio are out of sync, the video being much to slow. The sound in this page is with nellymoser compression.

Playing the downloaded .swf with ruffle-desktop, video and audio are synchronized.

Platform is openSuse Leap 15.2
Firefox version is 78.6.1esr (64-Bit)
ruffle-desktop and Firefox extension where both built from https://github.com/relrelb/ruffle/tree/b8543ecf77261ea9909a47b90f7ba5e76fa4550e, as the current nightly Firefox extension (2021-02-01) still doesn't support nellymoser.

### PR DESCRIPTION
**Pull request by Herschel**
This is a proof-of-concept first pass at syncing animation and audio playback when an embedded audio stream is playing. These streams are formed by `SoundStreamBlock` SWF tags and cause the Flash Player to skip frames in an attempt to keep the audio and video in sync.

 * Get the current time of the primary audio stream from the `AudioBackend`.
    * We consider the first "stream" sound to be the primary stream; typically only one stream is playing anyway.
 * Get the animation time of the clip that owns the audio stream.
   * Animation time is `(clip.current_frame() - stream.start_frame) / clip.movie().frame_rate`.
 * Compare audio time to animation time. If they differ by some threshold, adjust the playback speed to compensate.
   * Speed up or slow down the player by adding an offset to `Player::frame_accumulator`.
   * If they differ by a large threshold, we're way out of sync, so stop the stream completely. The movie clip will restart the sound at the correct position on the next frame.
 * Improve the handling of audio tracks with gaps (frames without `SoundStreamBlock` tags) on both desktop and web, particularly for MP3 tracks.
   * Specifically this will fix cases such as #3817 where a 1 or 2 frame gap should still play seamless audio. 
   * For MP3s, instead of stopping the stream when a SoundStreamBlock is missing, keep a running tally of the number of samples read from the stream vs. the number of timeline frames elapsed, and only create a gap if all of the samples have been consumed.
   * More research is needed as this isn't 100% accurate -- specifically the behavior of an MP3 stream track with larger gaps.
   * The Flash IDE seems to export and play larger Mp3 gaps out of sync in a strange way, for example: [stream_gaps_mp3.zip](https://github.com/ruffle-rs/ruffle/files/7807093/stream_gaps_mp3.zip).
   * But this Flash behavior is buggy-ish and was generally avoided in actual Flash content. (The trick was to fill gaps with a silent sound).

Here's a metronome SWF for testing:
[metronome.zip](https://github.com/ruffle-rs/ruffle/files/7358922/metronome.zip)

On Windows, you can test by clicking and holding on the window's maximize or close button, which will freeze the player until you drag off and release the mouse button. You can then see if the player recovers sync.
 
Fixes #3020, #3663, #3817, #3958.

### CODE REVIEW

**Comment by Toad06** [comment]
The PR looks good to me, it also fixes a bunch of issues in the early episodes of Steppenwolf.

I've just noticed a small regression that only happens on the Web targets:
- In [Steppenwolf 2-2](http://steppenwolf.epizy.com/steppenwolf/med/episode6.swf), start a new game.
- During the intro, you should hear "We've got to get away fast...", but the stream seems to start a bit too late, so the "we" is never heard. However, the sound still remains perfectly synced.

PS: Thanks for the amazing work and Happy New Year! 😉

**Comment by torokati44** [comment]
All the loops mentioned in https://github.com/ruffle-rs/ruffle/issues/3817#issuecomment-808728346 are playing back without stutter indeed!

**Comment by Herschel** [comment]
Going to merge as this is a nice improvement. Still more research that can be done here:
 * The behavior seems heavily influenced by the `samples_per_block` value in the SoundStreamHead tag. Specifically it seems that the player framerate gets clamped to the samples_per_block rate of the current stream (`stream.format.sample_rate / stream.samples_per_block`). It'd probably good to use this value directly for `frame_time` in the player's frame loop when a stream is playing.
 * I was hoping to simplify the code by applying the MP3 sample-counting logic to all sound types, but this didn't seem to be right in my initial tests. Even a single frame missing a SoundStreamBlock will cause an audible gap in a PCM sound, even if prior blocks had extra samples to compensate.

File: core/src/player.rs [hunk 1]
```diff
@@ -512,7 +512,8 @@ impl Player {
 
         if self.is_playing() {
             self.frame_accumulator += dt;
-            let frame_time = 1000.0 / self.frame_rate;
+            let frame_rate = self.frame_rate;
```
Comment (adrian17) [inline core/src/player.rs:515]: (unused change?)

File: core/src/backend/audio.rs [hunk 2]
```diff
@@ -438,6 +452,48 @@ impl<'gc> AudioManager<'gc> {
...
+            })
+            .next()
+            .unwrap_or_default();
+
+        if skew.abs() >= Self::STREAM_RESTART_THRESHOLD {
+            // Way out of sync, let's stop the entire stream.
+            // The movie clip will probably restart it naturally on the next frame.
+            let instance = &self.sounds[i];
+            audio.stop_sound(instance.instance);
+            self.sounds.swap_remove(i);
```
[the remark this answers is not in the document]
Comment (Herschel) [reply core/src/backend/audio.rs:486]: Nope, "stream" audio tracks are not observable by AS, and no event is fired when this occurs. One exception might be [`SoundMixer.computeSpectrum`](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/media/SoundMixer.html#computeSpectrum()) in AS3 to inspect the global audio output, but this is non-deterministic based on the state of the player anyway.

**Pull request merged**